### ISSUE TITLE
P3: Update Invited Unregistered 'Join Project' Workflow

### ISSUE THREAD

**Issue by ajefts** [appirio-tech/connect-app#3310]
**Objective:** Remove secondary step of pop-up window with 'Join Project/Cancel' that appears after the user has already clicked 'Join Project' in 'all projects' dashboard. 

<p></p>

**Requirements** - Workflow: 

[Imagining that users has already completed the steps of sending invitation/receiving invitation/registering (if necessary), and that they are now in Connect on the 'all projects' dashboard]

- User sees the project they have been invited to on their 'all projects' dashboard.

- User sees two action options: 'Join Project' as a primary button, or 'decline' listed in subscript underneath the 'Join Project' button.

  - **Option A:** User clicks 'Join Project' and they are immediately added to the project. 

  - **Option B:** User clicks 'Decline' and the project is removed from their 'all projects' dashboard.

<p></p>

In the workflow described above, the secondary reconfirmation of 'Join Project' pop-up window should not appear.

**Comment by ajefts** [comment]
Link to feature: https://tcprod.productboard.com/feature-board/planning/features/2761599

**Comment by RishiRajSahu** [comment]
@maxceem if this could be added to bug bash ?

**Comment by maxceem** [comment]
For testing the next project can be used http://local.topcoder-dev.com:3000/projects/8369. See [demo video](https://monosnap.com/file/cJFLHSZIX21Z8xESSYxuots1PHtFcD) on how I use user `pshah_manger` to invite user `pshah_customer` and after `pshah_customer` can see `Join project` button on the project listing page.

So as per task description above we should do the next things:
1. Add `decline` blue link under the `Join project` button on the project listing page (in both `grid` and `list` view).
2. When we click `Join Project` button we should redirected to the project details page, show loader indicator while we are sending request to the server to accept invitation. As soon as invitation is accepted we can start loading project. See how it's done when we click `Join Project` button in the dialog on the project details page.
3. When we click `Decline` we should hide the project in the project list, while sending request to the server to decline invitation. If error happens during invitation declining we should show red error alert in the bottom right corner (reuse Alert) with error message and show the project back in the project listing.

On the second thoughts I think we should implement this issue after https://github.com/appirio-tech/connect-app/issues/3311

**Comment by RishiRajSahu** [comment]
@maxceem all correct except - 
> When we click Join Project button we should redirected to the project details page,
we don't redirect user to project details page instead hide the Join/Decline button/link once the action is done.

do you think any conflicts will be there if we do #3311 / this issue together ?

**Comment by maxceem** [comment]
> do you think any conflicts will be there if we do #3311 / this issue together ?

No, I was thinking that we would redirect to the project details, so the current task would be partially implemented by #3311. But as per your update above, we don't have to redirect to the project details page, thus these two issues now can be done in parallel.

One more question though:
1. If we see a project with `join`/`decline` button in the project list and we click the link to the project we should still see the current join/cancel dialog, right?

    ![image](https://user-images.githubusercontent.com/146016/65386483-efab3400-dd44-11e9-8ba1-205447e07d74.png)

**Comment by RishiRajSahu** [comment]
@maxceem no actually we have to get rid of this dialog, but I got your point, and there are two scenarios coming to my mind now - 
1. User clicks on Join/decline button on listing page
2. User directly open the project for which is invited

Behaviour 1. - Keep user on listing page, don't show dialog, simply remove/fade-away the join/decline option. Can we some progress indicator there - like small loading icon ??

Behaviour 2 - show the dialog as we are showing right now

**Comment by maxceem** [comment]
Sum up.
Implement this task as per comment https://github.com/appirio-tech/connect-app/issues/3310#issuecomment-533863271

To show success and error alerts we can use the next place https://github.com/appirio-tech/connect-app/blob/dev/src/reducers/alerts.js#L1

For loading indicator we can use the next component https://github.com/appirio-tech/connect-app/blob/dev/src/components/LoadingIndicator/LoadingIndicator.jsx#L10 with `isSmall` prop.

**Comment by maxceem** [comment]
This issue is now open for pick up.

**Comment by PrakashDurlabhji** [comment]
@maxceem give it to me thank you

**Comment by PrakashDurlabhji** [comment]
@maxceem some more time

**Comment by PrakashDurlabhji** [comment]
@maxceem PR ready https://github.com/appirio-tech/connect-app/pull/3334

**Comment by maxceem** [comment]
@PrakashDurlabhji the PR which you provided doesn't implement any functionality which is required in this task. As we don't have much time and you still have another issue to complete which also goes too slowly I'm unassigning you from this issue so somebody else can work on it, while you work on the other issue.

**Comment by Gauravseta** [comment]
@maxceem please give me. I will complete it

**Comment by maxceem** [comment]
Sure @Gauravseta. But only 12 hours for this task as we don't have time. After 12 hours I will un-assign if not ready yet.

**Comment by Gauravseta** [comment]
@maxceem releasing it as wont be able to complete it

**Comment by rashmi73** [comment]
@maxceem pls give it to me

**Comment by rashmi73** [comment]
@maxceem PR is ready with minor changes left out while functionality is completed almost.
https://github.com/appirio-tech/connect-app/pull/3337

Confusion: 
1. **Add decline blue link under the Join project button**
any similar implementation anywhere in project, confused by the word "under" where exactly it needs to be placed? below?

2.**should show alert popup in right bottom corner You joined the project**
I guess it must be on left corner? as all similar popups are shown on left corner.e.g when attachements are attached succesfully. 


3.**bottom right corner (reuse Alert) with error message**
I guess again it should be on left corner? take my PR and you can login with pshah_customer and click on decline button to see red error having message. is that what required?

**Comment by maxceem** [comment]
Thanks for quick PR @rashmi73 

1. We don't have such example. I mean it should be below. So "Join Project" button and below blue link "Decline".

2. Yes, I mean standard **black** alert popup on the left.

3. Yes, I mean standard **red** alert popup on the left.

**Comment by rashmi73** [comment]
@maxceem Kindly check the screenshot below:

1. in projectSearch.projects[1] we have members and invites array through which we can find whether current user is member/invited or not(though it is not working perfectly, backend giving some wrong data even on decline/join project click)

2. I agree we can find whether user is invited or not from above logic, and our requirement says on clicking "decline" remove the project from project list visible. but how to find whether current user has rejected invitation? as we dont have any rejection array?

3. Also for one time i can achieve above requirement once, when user clicks on decline, i get sucess call back in promise and check whether it was decline then I can filter project from all projects, but that will be just once, because when I willl refresh the page then project will appear again, so I feel we need rejection array or different logic? If I am wrong can you please correct or guide me?

4. because user object present in invites array guarantees that user was invited, but absence from it doesnt guarantee that user has rejected it, because user may never have been invited too.

5. Or do I need to see "status" field inside invites array which may be "rejected" or "accepted" and by that I need to find it? is my direction correct?

![3310](https://user-images.githubusercontent.com/41687423/65638263-5d848380-e003-11e9-8ba9-5b86ac0492fb.png)

**Comment by maxceem** [comment]
3. > Also for one time i can achieve above requirement once, when user clicks on decline, i get sucess call back in promise and check whether it was decline then I can filter project from all projects, but that will be just once, because when I willl refresh the page then project will appear again, so I feel we need rejection array or different logic? If I am wrong can you please correct or guide me?

     This is correct way to go. Though I see where confusion comes from. If you invite user `pshah_copilot` or `pshah_customer` and after reject. Then project has to disappear from the list after reload, because such users cannot see the project they are not invited to.

       Though if you invite a powerful user, like `pshah_manger`, then if such user declines the invitation, after reload he can still see the project even he is not invited. So let's do the next thing. For users who has [manager or admin roles](https://github.com/topcoder-platform/tc-project-service/blob/dev/src/constants.js#L47-L54) after decline invitation we just remove the buttons, because they can still see the project in the list.
For other users we hide the project. After refresh such users shouldn't see the project on the list. If you can see it, then it must be an issue on the Project Service. Please, let me know if you see such issue.

Regarding your questions 1, 2, 4 and 5. We don't have to know in general if user rejected invitation before. If user rejected invitation then he doesn't have invitation anymore, so there wouldn't be buttons. We just have to hide buttons or project after user clicks any of buttons and request to server is successful. So we should do it as you described in 3.

@rashmi73 Let me know if I miss something or you have other concerns.

**Comment by rashmi73** [comment]
@maxceem yes I do have other concerns

**For users who has manager or admin roles after decline invitation we just remove the buttons, because they can still see the project in the list**

I need something from backend always such as "rejected" value in status field of invites array then only I would be able to represent the flow by checking if manager and rejected then dont show buttons. in another case if not manager and rejected then dont show project itself.
To achieve this need Project service to function properly which is currently not. hence as you mentioned there is an issue in project service.

**Comment by maxceem** [comment]
@rashmi73 hmm, for me backend already sends the correct status of invitation.

1. I click `Decline` button.
2. I see that we send the next request:

    ![image](https://user-images.githubusercontent.com/146016/65830625-b003bf80-e2e3-11e9-8905-cccaa92e8e16.png)

3. And response from the server looks like this, with `refused` status:

     ![image](https://user-images.githubusercontent.com/146016/65830638-c7db4380-e2e3-11e9-8051-500f25d1f1a3.png)

4. If I click `Join project` button.
5. I see that we send the next request:

    ![image](https://user-images.githubusercontent.com/146016/65830663-0b35b200-e2e4-11e9-983b-42a29f4c551a.png)

6. And response from the server looks like this, with `accepted` status:

    ![image](https://user-images.githubusercontent.com/146016/65830666-1a1c6480-e2e4-11e9-91ec-e6ee5f5ba5e3.png)

Do you have some steps to reproduce when server returns wrong status? Note: I use user `pshah_manger` to invite user `pshah_copilot` for testing. I use `Customer team` dialog to invite users.

**Comment by RishiRajSahu** [comment]
@maxceem yep project service seems to be sending the correct data, @rashmi73 can't we check `if manager and PENDING` instead of `if manager and rejected` to show/hide button ?

**Comment by rashmi73** [comment]
@maxceem @RishiRajSahu I guess you have misunderstood what I am trying to convey.

1. the screenshot u pasted above shows correct response is received for that particular http call.

2. But I am talking regarding the list which is fetched. on fetching list of projects, in "ProjectSearch.projects[1].invites[1].status"   I shall get status: "rejected"/"accepted" . but at this place 
I get pending always. hence how to determine then? and this is only way I can apply logic concretely. because on hitting join/reject button I may be able to do manipulation thereitself. then what if user refreshes browser? then I need the values to be reflected in invites array as well.

I hope I am able to make you understand.

**Comment by maxceem** [comment]
Thanks for detailed questions @rashmi73.

> then what if user refreshes browser?

When user refreshes browser, rejected or accepted invitations should be NOT in the list. As soon as we reject or accept invitation it's not anymore returned by the server. As we don't have such rejected or accepted invitations in the list we don't show buttons anymore, which is what we need. 

If you still see invitation in the list with the `pending` status after you rejected or accepted them then it's an issue on the server. Let me know if there are any steps you can reproduce it.

**Comment by rashmi73** [comment]
@maxceem commented on PR

**Comment by rashmi73** [comment]
@maxceem online for 4 hours?

**Comment by maxceem** [comment]
@rashmi73 no, I'll be online tomorrow around 10 am GMT+8

**Comment by rashmi73** [comment]
@maxceem working

**Comment by maxceem** [comment]
@rashmi73 As there is still no solution I have to un-assign you from this issue. The desired time for solving issues in our Bug Bash is 12 hours, but this issue took too long, and unfortunately, we cannot wait any longer as this issue is already delayed for release.

This issue is no longer in Bug Bash.

**Comment by RishiRajSahu** [comment]
Pushed to prod.
